mdagent/tools/base_tools/analysis_tools/thermo_tools.py

```diff
@@ -0,0 +1,335 @@
+        charge_json = path_registry.get_mapped_path(charge_file_id)
+    except Exception:
+        raise FileNotFoundError(
+            "Failed to load charge file. File ID not found in path registry."
+        )
+    try:
+        charges = load_charge_json(charge_json)
+    except Exception:
+        raise Exception(
+            "Failed to load charge file. Charge file should be a "
+            "json file mapping each atom in the trajectory to its partial charge."
+        )
+    try:
+        charges = match_charges_to_traj(charges, traj)
+    except Exception:
+        raise Exception(
+            "Failed to match charges to trajectory. Please ensure that "
+            "Each atom in the trajectory is mapped to its partial charge "
+            "in the json file."
+        )
+
+
+class ComputeDipoleMoments(BaseTool):
+    name = "ComputeDipoleMoments"
+    description = """Compute the total dipole moment for each frame in a
+    molecular dynamics trajectory. Requires a trajectory file and a json file mapping
+    each atom in the trajectory to its partial charge.
+    Optionally requires a topology file.
+    Returns an array of dipole moments for each frame of the trajectory,
+    written to a file."""
+    path_registry: PathRegistry = PathRegistry().get_instance()
+
+    def __init__(self, path_registry: PathRegistry | None = None):
+        super().__init__()
+        if path_registry is not None:
+            self.path_registry = path_registry
+
+    def _compute_dipole_moments(self, traj: md.Trajectory, charges: np.ndarray):
+        return md.dipole_moments(traj, charges)
+
+    def _save_dipoles_to_file(self, dipole_moments: np.ndarray, traj_file: str):
+        file_name = self.path_registry.write_file_name(
+            type=FileType.RECORD,
+            record_type="dipole_moments",
+            file_format="csv",
+        )
+
+        save_path = f"{self.path_registry.ckpt_files}/{file_name}"
+        description = (
+            f"Dipole moments for each frame in the trajectory for file {traj_file}"
+        )
+
+        file_id = self.path_registry.get_fileid(
+            file_name=file_name, type=FileType.UNKNOWN
+        )
+        self.path_registry.map_path(file_id, save_path, description)
+
+        np.savetxt(
+            save_path,
+            dipole_moments,
+            delimiter=",",
+            header="Dipole_X,Dipole_Y,Dipole_Z",
+            comments="",
+        )
+        return file_id
+
+    def _run(self, traj_file: str, charge_file: str, top_file: str | None = None):
+        try:
+            traj = load_single_traj(
+                path_registry=self.path_registry,
+                traj_fileid=traj_file,
+                top_fileid=top_file,
+            )
+            if not traj:
+                raise Exception("Failed to load trajectory file.")
+        except Exception as e:
+            return str(e)
+
+        try:
+            charges = get_charges(self.path_registry, traj, charge_file)
```

Comment:
How is the agent getting or making the charge_file prior to using this tool?

Reply:
i think in the case of a vector, it would make sense to also capture descriptors such as the magnitude and direction w.r.t. to the z-axis (?) or something. 

And even though I think I've said this a lot (sorry!) maybe the magnitude of the dipole per frame and or its direction can be cool figures

Reply:
ah, the user can provide it. I added another function though that can try to load the traj into a system and get charges using openmm if not provided